### DISCUSSION TITLE
Preset for long-read RNAseq with cell barcode

### DISCUSSION THREAD

**Discussion by Euchiz** [milaboratory/mixcr#1681] [Q&A] [answered]
Hi everyone,

I am currently working with single-cell Oxford Nanopore RNAseq data that contains both a 10x cell barcode and a UMI at the beginning of each read. Unfortunately, there seem to be no existing presets (or could anyone point out which should I use) that cater to this specific combination of single-cell barcoding and long-read sequencing parameters.

While I have reviewed the existing long-read RNAseq presets from the website, they do not include options for handling single-cell barcodes. I attempted to modify the YAML configuration files to create a custom preset that combines the 10x single-cell workflow with long-read sequencing parameters. 

I have reviewed a previous discussion on a related topic ([Discussion #1024](https://github.com/milaboratory/mixcr/discussions/1024)), but the provided guidance was based on outdated YAML configurations.

Could you please provide assistance or guidance on creating a custom preset for this specific use case? 
FYI here is the read structure I am working with:

```
"^(CELL:N{16})(UMI:N{12})(R1:*)"
```
Any help with the correct configuration or a new preset would be greatly appreciated.

By the way, I have already used ([wf-single-cell](https://github.com/epi2me-labs/wf-single-cell)) to refine my nanopore reads. Does it mean I can remove the refinement step and use the result bam file from wf-single-cell directly for alignment and assembling?

Thank you for your support and for developing such a powerful tool!

**Comment by mizraelson** [comment, marked as answer]
Hi,
I assume you use 10xV3 kit, correct? Are you sure the reads do not contain any ONT adapter sequences before the cell barcode at the beginning? Assuming the pattern can be in reverse complement, as is usually the case with ONT, I would try the following custom [preset](https://github.com/user-attachments/files/15597384/preset.zip). Unzip the YAML file and put it the directory where you run MiXCR from or to the `~/.mixcr/presets` folder. The command would then be:

```
mixcr analyze local:10x-ont \
    --species hsa \
    input_R1.fastq.gz \
    input_R1.fastq.gz \
    output
```
Use [the latest develop MiXCR version](https://link.milaboratory.com/software/mixcr/mixcr-develop.zip).

I recommend using the raw dataset, not preprocessed by `wf-single-cell`.

**Reply by Euchiz** [reply]
Thank you so much! You are right there should be adapter before the barcode. I changed the expression and it worked.

**Reply by mizraelson** [reply]
Great! Maybe you can share the adjusted preset (with the adapter) here in case other users struggling with the same protocol.

**Reply by zztin** [reply]
Hi @mizraelson , 

I am following the example you provided above with the custom preset provided. However, the MiXCR complaints that "No preset with name "groupClones-base".

Do you know how to fix this?


Also, do you have more information about the adapter @Euchiz  mentioned? Any tips how I can modify my tag_pattern with the adapter sequences? 

Thanks a bunch!

**Reply by mizraelson** [reply]
The groupClones-base should be changed to assembleCells-base. You would also need to add the following parameter:
--assemble-contigs-by VDJRegion with the mixcr analyze command.

I am not sure what pattern @Euchiz ended up using, but the idea is to “trim” the ONT adapter sequence before the CELL barcode starts and after the C gene sequence ends. It should look something like this:

^*atgcatgcatgcatgc(CELL:N{16})(UMI:N{12})(R1:*)atgcatgcatgcatgc*

However, the sequences should match the ones you actually have in your reads.

**Reply by zztin** [reply]
Hi @mizraelson, 
Thank you so much for the answer. I modified the last 2 lines as you suggested. 

```
groupClones:
  inheritFrom: groupClones-base
```
to 
```
groupClones:
  inheritFrom: assembleCells-base
```
and alternatively
```
assembleCells:
  inheritFrom: assembleCells-base
```


Both give a `NullPointerException`. See *error traceback* below. Do you have some suggestions?

*error traceback*: 
```Please copy the following information along with the stacktrace:
   Version: 4.7.0; built=Wed Aug 07 21:19:48 CEST 2024; rev=976ba14139; lib=repseqio.v5.1
        OS: Linux
      Java: 23.0.1-internal
  Abs path: /hpc/pmc_holstege/lchen/VDJ
  Cmd args: analyze local:10x-ont-modified --species hsa --tag-pattern ^.*TTTTCCTACACGACGCTCTTCCGATCT(CELL:N{16})(UMI:N{12})(R1:*)CCCATGTACTCTGCGTTGATACCACTGCTT /hpc/pmc_holstege/lchen/data/ONT_raw/LX752/run_1095/LX752_S1095_hac.fastq.gz ./output/LX752_ONT_test_generic
picocli.CommandLine$ExecutionException: Error while running command analyze java.lang.NullPointerException:  must not be null
        at com.milaboratory.mixcr.cli.Main.registerExceptionHandlers$lambda-17(SourceFile:420)
        at picocli.CommandLine.execute(CommandLine.java:2088)
        at com.milaboratory.mixcr.cli.Main.execute(SourceFile:105)
        at com.milaboratory.mixcr.cli.Main.main(SourceFile:101)
Caused by: java.lang.NullPointerException:  must not be null
        at com.milaboratory.mixcr.cli.CommandAssemble.validateParams(SourceFile:564)
        at com.milaboratory.mixcr.cli.MiXCRParamsResolver.validateBundle(SourceFile:60)
        at com.milaboratory.mixcr.cli.MiXCRParamsResolver.validateBundle(SourceFile:38)
        at com.milaboratory.o.w.resolve(SourceFile:180)
        at com.milaboratory.o.w.resolve$default(SourceFile:171)
        at com.milaboratory.mixcr.cli.CommandAnalyze$Cmd.run0(SourceFile:341)
        at com.milaboratory.mixcr.cli.MiXCRCommand.run(SourceFile:37)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at com.milaboratory.mixcr.cli.Main.registerLogger$lambda-32(SourceFile:539)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        ... 2 more
```

**Reply by zztin** [reply]
Since I am new to mixcr and I think it's best to read documentations about how to build custom preset and what parameters are available. However, I couldn't really find that info yet. Can you point me to the documentation?

For example, you suggested to add "--assemble-contigs-by VDJRegion", I added it in the command line. Can you also provide an example how to add this in the customized preset? 

Lastly, for the adapter sequence `^atgcatgcatgcatgc(CELL:N{16})(UMI:N{12})(R1:)atgcatgcatgcatgc*`
I figured that the start of a Nanopore read is not constant. I now created this tag-pattern:
`^*TTTTCCTACACGACGCTCTTCCGATCT(CELL:N{16})(UMI:N{12})(R1:*)CCCATGTACTCTGCGTTGATACCACTGCTT*`

In the beginning of the read preceeding `TTTTCCTT` there is ~50bp sequence. Can I use a asterik to represent them? I also want to know if this tag-pattern is used as exact matches or is there any kind of fuzzy match? Should I provide a full adapter sequence given that there will be likely no exact match due to the high error rate of ONT reads? 
And, do you suggest to remove the pattern `^*` altogether in this case if I only provide the suffix of the adapter not the complete adapter? 

Thank you very much for your input.

**Reply by mizraelson** [reply]
We do not yet have documentation for building a custom preset, unfortunately.

For `--assemble-contigs-by VDJRegion`, you can add the following to the mixins:

```
mixins:
  - type: SetTagPattern
    tagPattern: ^(CELL:N{16})(UMI:N{12})(R1:*)
  - type: MaterialTypeRNA
  - type: SetContigAssemblingFeatures
     features: VDJRegion
```

You can use an asterisk (`*`) to match any number of unknown nucleotides. Capital letters represent an exact match (e.g., TTTTCCTT), while lowercase letters indicate a fuzzy match (e.g., ttttcctt).

You may want to provide a shorter adapter sequence to account for sequencing error rates, but it should still be long enough to anchor the pattern. `^*` should remain in the pattern— `^` ensures the pattern starts at the beginning, and `*` allows for any number of nucleotides before the anchor sequence.

**Reply by zztin** [reply]
Hi @mizraelson , 

Thank you for your response. You might missed the previous question in the same thread. Can you also take a look at give some advice on this one?
https://github.com/milaboratory/mixcr/discussions/1681#discussioncomment-12424336

Your input is very appreciated!

**Reply by mizraelson** [reply]
Yes, sorry about that. The correct syntax is:

```
assembleCells:
    inheritFrom: assembleCells-base
```

Can you share the preset file that you run. You can send it to support@milaboratories.com

**Reply by zztin** [reply]
Hi, I share the preset to the email address you mentioned above and also here. (unzip to get the yaml file)
Can you please try running it at your end to test? 
Let me know if you want to have a test dataset.
[10x-ont-modified.yaml.zip](https://github.com/user-attachments/files/19184463/10x-ont-modified.yaml.zip)

**Reply by acherman** [reply]
Chiming in to +1 @zztin null pointer issue. My .yaml is identical except for the adapter sequence.

**Reply by mizraelson** [reply]
The assembling feature was missing. We had a default CDR3 value in the previous version but now it has to be specified explicitly. Please find the corrected version attached.
[preset.zip](https://github.com/user-attachments/files/19200753/preset.zip)

**Reply by acherman** [reply]
Thank you! Got it to run successfully but I realized that my [5' v3 10x needs the 3M-5pgex-jan-2023 barcode list](https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-inclusion-list-formerly-barcode-whitelist). I gave it a shot and got

`picocli.CommandLine$ExecutionException: Error while running comma
nd refineTagsAndSort java.lang.IllegalArgumentException: Unknown 
sequence set name: "3M-5pgex-jan-2023"`

Is that barcode list unavailable or does it have a different name? Also, let me know if I should move this to a separate discussion!

**Reply by acherman** [reply]
Also, from a practical standpoint, if I'm really just interested in the clonotypes present in the data (irrespective of which cell they come from) is there any reason that the `generic-ont` preset would give misleading results? I've run my data with that preset and got apparently sensible results.

**Reply by mizraelson** [reply]
Yes, this is a new whitelist. I will add it in the next update.

Also, if you don’t need to distinguish individual cells, you can use the `generic-ont` preset. However, `generic-ont-with-umi` might be a better option since it allows you to use UMIs for error correction.

**Reply by acherman** [reply]
Thanks so much for all your help @mizraelson! I've learned a lot from the issues and discussions sections here.

**Comment by JackYGhannam** [comment]
Hello! 

Thank you @mizraelson for all of your hard work on this. A follow-up question: would this custom preset also work if one wanted to process 10x 5'v2 single-cell data for which TCRs were subsequently enriched and sequenced via nanopore sequencing?  Thanks again!

**Reply by mizraelson** [reply]
Yes, but you should change the pattern and the cell barcodes whitelist accordingly to `^(CELL:N{16})(UMI:N{10})\^(R2:*)` and `builtin:737K-august-2016` in the preset file.

**Comment by austinc26** [comment]
HI!

Thanks @mizraelson for helpful insight on this! I was just wondering if I have a 10x Multiome dataset that was later sequenced using Oxford Nanopore, could I use this preset if I trimmed the TSO and R1 adapters and discarded reads that did not have either adapter and also used the 737k-arc-v1 cell barcode whitelist in the yaml file. Thanks!

**Comment by mizraelson** [comment]
I am not sure I understand the data structure you will get. But the `737k-arc-v1` barcodes list is not build into MiXCR because it is not used for VDJ libraries. We might add it though in the upcoming version.
